File: app/src/main/java/pl/laskowski/marcin/type/ClassicRectangle.java
package pl.laskowski.marcin.type;

import pl.laskowski.marcin.model.Sudoku;
import pl.laskowski.marcin.model.dividers.*;

/**
 * Created by Marcin Laskowski.
 */

public class ClassicRectangle extends SudokuVariant {

    private final int blockWidth;
    private final int blockHeight;

    public ClassicRectangle(int width, int height, int blockWidth, int blockHeight) {
        super(width, height);
        this.blockWidth = blockWidth;
        this.blockHeight = blockHeight;
    }

    @Override
    public int regionSize() {
        return blockWidth * blockHeight;
    }

    @Override
    protected SudokuDivider divider() {
        return new ComplexDivider(
                new RowDivider(),
                new ColumnDivider(),
                new BlockDivider(blockWidth, blockHeight));
    }

    @Override
    public Sudoku template() {
        return Sudoku.blank(width(), height());
    }

    public int getBlockHeight() {
        return blockHeight;
    }

    public int getBlockWidth() {
        return blockWidth;
    }


}

File: app/src/main/java/pl/laskowski/marcin/type/ClassicRectangle.kt
package pl.laskowski.marcin.type

import pl.laskowski.marcin.model.dividers.SudokuDivider
import pl.laskowski.marcin.model.dividers.ComplexDivider
import pl.laskowski.marcin.model.dividers.RowDivider
import pl.laskowski.marcin.model.dividers.ColumnDivider
import pl.laskowski.marcin.model.dividers.BlockDivider
import pl.laskowski.marcin.model.Sudoku

open class ClassicRectangle(
    width: Int,
    height: Int,
    val blockWidth: Int,
    val blockHeight: Int
) : SudokuVariant(width, height) {

    override fun regionSize(): Int {
        return blockWidth * blockHeight
    }

    override fun divider(): SudokuDivider {
        return ComplexDivider(
            RowDivider(),
            ColumnDivider(),
            BlockDivider(blockWidth, blockHeight)
        )
    }

    override fun template(): Sudoku {
        return Sudoku(this)
    }
}